spring-web/src/main/java/org/springframework/web/service/invoker/RequestParamArgumentResolver.java

```diff
@@ -54,18 +58,62 @@
  */
 public class RequestParamArgumentResolver extends AbstractNamedValueArgumentResolver {
 
+	private boolean formatAsSingleValue;
 
 	public RequestParamArgumentResolver(ConversionService conversionService) {
 		super(conversionService);
+		this.formatAsSingleValue = true;
+	}
+
+	public RequestParamArgumentResolver(ConversionService conversionService, boolean formatAsSingleValue) {
+		super(conversionService);
+		this.formatAsSingleValue = formatAsSingleValue;
 	}
 
 
 	@Override
 	@Nullable
 	protected NamedValueInfo createNamedValueInfo(MethodParameter parameter) {
 		RequestParam annot = parameter.getParameterAnnotation(RequestParam.class);
+		boolean formatAsSingleValue = this.formatAsSingleValue;
+		if (annot == null) {
+			return null;
+		}
+
+		List<String> invalidContentType = List.of(
+				MediaType.APPLICATION_FORM_URLENCODED_VALUE.toLowerCase(),
+				MediaType.MULTIPART_FORM_DATA_VALUE.toLowerCase(),
+				MediaType.MULTIPART_FORM_DATA_VALUE.toLowerCase(),
+				MediaType.MULTIPART_MIXED_VALUE.toLowerCase(),
+				MediaType.MULTIPART_RELATED_VALUE.toLowerCase());
```

Comment:
I would replace this with a protected method that takes the same input and decides whether to format as single value or not. The method can return `true` if `formatAsSingleValue` is set except when the content type is form data or multipart data. For the multipart check, you could use `contentType.getType().equals("multipart")`. 

The protected method would allow a subclass to further refine the decision. It would be a good idea to expose the formatAsSingleValue property with a getter for subclasses to be able to check. I think the property could also have a setter rather than being a constructor argument.

Reply:
On https://github.com/spring-projects/spring-framework/commit/d4f666bd318749a05f06ca700daad640b57ea685 commit, I add new protected method `isMultiValueFormContentType(MediaType contentType)` and use it on new override method you made 🙂